src/python/build_management/build_manager.py

```diff
@@ -738,8 +739,21 @@ class FuchsiaBuild(RegularBuild):
   FUCHSIA_BUILD_REL_PATH = os.path.join('build', 'out', 'default')
   FUCHSIA_DIR_REL_PATH = 'build'
 
+  def __init__(self, *args, **kwargs):
+    super().__init__(*args, **kwargs)
+
+    # Decide per-build whether to use undercoat, based on rollout level
+    rollout_level = environment.get_value('FUCHSIA_UNDERCOAT_ROLLOUT_LEVEL', 0)
+    self.use_undercoat = random.random() < rollout_level
+    logs.log('Undercoat enabled: %s' % self.use_undercoat)
+
   def _get_fuzz_targets_from_dir(self, build_dir):
     """Overridden to get targets list from fuchsia."""
+    if self.use_undercoat:
+      # list_fuzzers requires a running instance, so we can't list these until
+      # later in setup() below
+      return []
```

Comment:
Taking a closer look here it's a rather difficult to untie the target selection logic from the unpacking process.

Something like this is reasonable, but will result in a log_error every setup() because we return this empty list.

Here's a rough approach that I can think of:

Add a `_pick_fuzz_target` **method** to the `Build` class which just calls the top level `_set_random_fuzz_target_for_fuzzing_if_needed` by default. Then replace all the existing calls where appropriate to call the class method instead.  We still need the top-level function because it's called from outside the context of a Build as well. 

Then for fuchsia, we can just override `_pick_fuzz_target` to **always** be a no-op.

Then in `setup()` we can decide whether or not to use undercoat after the unpacking, and call a custom `_pick_fuzz_target_for_fuchsia` of some sort in both cases.

Does that make sense/seem reasonable?

Reply:
Thanks for the detailed suggestions! I did something similar to what you described, with a little bit of modification: mostly a) leaving out the extra layer of `_pick_fuzz_target_for_fuchsia` because it would simply be another bare wrapper to `_set_random_fuzz_target_for_fuzzing_if_needed` and b) it seemed slightly cleaner for `_pick_fuzz_target` to be an instance method instead of a class method, even though it doesn't need to be.  Happy to change those if you'd prefer otherwise, though.  I also split the setup function up a bit more in the process of implementing fallback logic, which hopefully makes the distinction between the two codepaths (legacy/undercoat) slightly clearer.